EIPS/eip-2566.md

```diff
@@ -0,0 +1,128 @@
+
+Example
+// Request
+curl -X POST --data '{"jsonrpc":"2.0","method":"eth_sendTransactionToContractFunction","params":[{see above}],"id":1}'
+
+// Result
+{
+  "id":1,
+  "jsonrpc": "2.0",
+  "result": "0xe670ec64341771606e55d6b4ca35a1a6b75ee3d5145a99d05921026d1527331"
+}
+
+## Rationale
+With backwards compatibility in mind, this EIP proposes augmenting the set of Ethereum RPC methods with an additional method instead of mutating the existing method. Precedent for adding a new RPC method comes from [EIP 712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md) in which adding the method `eth_signTypedData` is proposed for confirmation prompt security. As an alternate approach, the `eth_sendTransaction` method could be changed to accept its `data` argument in a human readable format, but this would break all existing code attempting to execute a contract function. Breaking this backwards compatibility would be a far bigger issue than the readability of an `data` argument so this EIP does not recommended this approach.
+
+This EIP's proposed `eth_sendTransactionToContractFunction` method is intended to parallel `eth_sendTransaction` as much as possible since both methods result in the same behaviour when executing a contract function. The `abi` field is an element of the contract's ABI that corresponds to the intended function. The `data` field is a non ABI-encoded array of arguments to be passed to the contract function. Dapps need not ABI-encode arguments as the data types of these arguments can be inferred from the `abi` field on the ProviderWallet side and encoded at that time.
+
+## Backwards Compatibility
+This EIP does not break backwards compatibility and addresses that point above.
+
+## Implementation
+The `data` field from `eth_sendTransaction` can be reconstructed in its non-human readable format from the `abi` and `data` fields in `eth_sendTransactionToContractFunction`. This allows implementations to broadcast a transaction with data in the correct format for the Ethereum virtual machine. The function signature can be built and hashed from the function ABI, and the arguments can be ABI-encoded according to the data types specified in the function ABI.
+
+## Security Considerations
+Displaying the contract address, function name, and argument name/values can provide additional security to users but it is not a guarantee that a function will execute as the user expects. A poorly implemented contract can still name its function `transfer` and accept `address` and `uint256` arguments - but there is nothing short of contract examination that will let a user know that this contract is indeed a valid ERC20 contract. This EIP does not intend to solve the larger problem around trust in a contract's code, but instead intends to give users better tools to understand exactly what is contained within the data they are broadcasting to the Ethereum network.
```

Comment:
Function selectors are only 4 bytes which means it is pretty trivial to create collisions for them.  Also, parameter names are not validatable at all so a malicious application can trivially rename the variables to whatever they want.

Consider checking out and championing something like https://github.com/ethereum/EIPs/issues/719 if you want to introduce a mechanism for informed consent that doesn't require trusting the application.

Reply:
@MicahZoltu Trying to think up an example for your collision comment. Are you suggesting that a contract could have a function with signature `doMaliciousThingXYZ123(address,unit256)` where the name `doMaliciousThingXYZ123` is constructed in such a way that the first 4 bytes of this signature hash could be the same as something else like `doNonMaliciousThing`. Then the malicious Dapp can provide `doNonMaliciousThing` in the ABI and trick the end user to actually execute `doMaliciousThingXYZ123`?

Reply:
If this scenario is what you are suggesting - I think that I was approaching the problem from this perspective: we can trust contracts since their source code is public and auditable. We cannot trust dapps (I'm using the word here to mean websites that interact with the chain) since most of their source code is hidden from users. For example, picture a sketchy dapp that is interacting with a well known and audited contract such as OMG ERC20. We can trust that the contract doesn't have any functions like `doMaliciousThingXYZ123(address,unit256)` since it's a public and audited contract. However, we cannot trust that the dapp is actually sending tokens to who the user intends to send them to. They could be swapping around addresses as in my example in the Motivation section. I feel like it's rare that someone examines the `data` field.

Reply:
Cheers on linking #719 I have been aware of your work there for a while. I will give it a closer examination.